From 55cc3979c3f59e75548e8652f37e889ccd45416c Mon Sep 17 00:00:00 2001
From: Mohsin Kazmi <sykazmi@cisco.com>
Date: Thu, 17 Jun 2021 14:57:01 +0200
Subject: [PATCH] net: prepare the outer ipv4 hdr for checksum

Preparation the headers for the hardware offload
misses the outer ipv4 checksum offload.
It results in bad checksum computed by hardware NIC.

This patch fixes the issue by setting the outer ipv4
checksum field to 0.

Signed-off-by: Mohsin Kazmi <sykazmi@cisco.com>
---
 lib/net/rte_net.h | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/lib/net/rte_net.h b/lib/net/rte_net.h
index 434435ffa2..48093830a0 100644
--- a/lib/net/rte_net.h
+++ b/lib/net/rte_net.h
@@ -128,8 +128,18 @@ rte_net_intel_cksum_flags_prepare(struct rte_mbuf *m, uint64_t ol_flags)
 	if (!(ol_flags & (PKT_TX_IP_CKSUM | PKT_TX_L4_MASK | PKT_TX_TCP_SEG)))
 		return 0;
 
-	if (ol_flags & (PKT_TX_OUTER_IPV4 | PKT_TX_OUTER_IPV6))
+	if (ol_flags & (PKT_TX_OUTER_IPV4 | PKT_TX_OUTER_IPV6)) {
 		inner_l3_offset += m->outer_l2_len + m->outer_l3_len;
+                /*
+                 * prepare outer ipv4 header checksum by setting it to 0,
+                 * in order to be computed by hardware NICs.
+                 */
+                if (ol_flags & PKT_TX_OUTER_IP_CKSUM) {
+                        ipv4_hdr = rte_pktmbuf_mtod_offset(m, struct rte_ipv4_hdr *,
+                                   m->outer_l2_len);
+                         ipv4_hdr->hdr_checksum = 0;
+                }
+        }
 
 	/*
 	 * Check if headers are fragmented.
-- 
2.17.1

